### DISCUSSION TITLE
can someone explain about the foreground feature?

### DISCUSSION THREAD

**Discussion by abhishek-m-raj** [781flyingdutchman/background_downloader#361] [Q&A] [answered]
what does enabling foreground download do in devices?
does enabling foreground means no background download?

i use redirect links for downloads and it does not start from where it left off when i close the app and reopen it
@781flyingdutchman

**Comment by 781flyingdutchman** [comment, marked as answer]
Enabling foreground is a technical thing on Android, where - if set - the WorkManager (the service that schedules background tasks) will start a task within an Android foreground service instead of a background service.  This has one advantage and several disadvantages, so should not generally be used. The advantage (and the only reason really to use it) is that it removes the 9 minute time limit on task completion, so if you have really big downloads or a very poor network *and* the task is not pausable, then you may consider activating it (this is rare). Android doesn't like you running foreground services, so the disadvantages are:
* Resource Consumption: Foreground services have a higher priority than background tasks, meaning they are less likely to be killed by the system when resources are scarce. However, this also means they consume more battery and memory, potentially impacting the device's performance and user experience.
* User Experience: Foreground services require a persistent notification in the notification tray to inform the user that the task is running. This can be disruptive and annoying for users, especially if they are not actively engaged with the app. Additionally, foreground services can prevent the device from entering a low-power state, further impacting battery life.

In general, it's recommended to use foreground services only for tasks that are critical and require immediate execution, even if the app is in the background. For most background tasks (including most download tasks, WorkManager's standard background execution mechanisms are sufficient and provide a better balance between resource consumption and user experience.

**Reply by abhishek-m-raj** [reply]
thanks i have another problem i addressed the downloads which use redirect links will not resume when app is restarted but they can be paused and resumed when app is in use

**Comment by DevUser611** [comment]
How this task completion time limit actually calculated ? Is it calculated from moment when you register first task or for each task separately ?
For Example if you have 800 download tasks and to complete all of them it takes 30 minutes but to complete 1 of the task it takes 2 seconds. Will all tasks be killed after 9 minutes (and therefore tasks that didn't event start) or not ?

**Comment by 781flyingdutchman** [comment]
It is per task, and measured from the moment a task starts (i.e. emits a `running` status). If a task exceeds the 9 minute run time, it is cancelled by the OS, but handled by the background downloader as failed (and potentially paused and resumed as a new task). Other tasks are not affected.

**Reply by DevUser611** [reply]
So i enqueue 800 task it will work fine ? Even when app goes to background
